packages/streaming-image-volume-loader/src/helpers/sortImageIdsAndGetSpacing.ts

```diff
@@ -45,28 +55,80 @@ export default function sortImageIdsAndGetSpacing(
       imagePositionPatient
     );
 
-    const distance = vec3.dot(positionVector, scanAxisNormal);
+    return vec3.dot(positionVector, scanAxisNormal);
+  }
+
+  /**
+   * If we have all image metadata, then sort by image position in 3D space, and
+   * calculate average slice spacing from the entire volume. If not, then use
+   * the sampled images (1st and middle) to calculate slice spacing, and use
+   * the provided imageId order. Correct sorting must be done ahead of time.
+   */
+  if (haveAllMetadata) {
+    const distanceImagePairs = imageIds.map((imageId) => {
+      const distance = getDistance(imageId);
 
-    return {
-      distance,
-      imageId,
-    };
-  });
+      return {
+        distance,
+        imageId,
+      };
+    });
 
-  distanceImagePairs.sort((a, b) => b.distance - a.distance);
+    distanceImagePairs.sort((a, b) => b.distance - a.distance);
 
-  const sortedImageIds = distanceImagePairs.map((a) => a.imageId);
-  const numImages = distanceImagePairs.length;
+    sortedImageIds = distanceImagePairs.map((a) => a.imageId);
+    const numImages = distanceImagePairs.length;
+
+    // Calculated average spacing.
+    // We would need to resample if these are not similar.
+    // It should be up to the host app to do this if it needed to.
+    zSpacing =
+      Math.abs(
+        distanceImagePairs[numImages - 1].distance -
+          distanceImagePairs[0].distance
+      ) /
+      (numImages - 1);
+  } else {
+    // Assume initial imageId array order is pre-sorted, but check orientation.
+    const prefetchedImageIds = [
+      imageIds[0],
+      imageIds[Math.floor(imageIds.length / 2)],
+    ];
+    sortedImageIds = imageIds;
+    const firstImageDistance = getDistance(prefetchedImageIds[0]);
+    const middleImageDistance = getDistance(prefetchedImageIds[1]);
+    if (firstImageDistance - middleImageDistance < 0) {
+      sortedImageIds.reverse();
+    }
+
+    // Calculate average spacing between the first and middle prefetched images,
+    // otherwise fall back to DICOM `spacingBetweenSlices`
+    const metadataForMiddleImage = metaData.get(
+      'imagePlaneModule',
+      prefetchedImageIds[1]
+    );
+    if (metadataForMiddleImage) {
+      const positionVector = vec3.create();
 
-  // Calculated average spacing.
-  // We would need to resample if these are not similar.
-  // It should be up to the host app to do this if it needed to.
-  const zSpacing =
-    Math.abs(
-      distanceImagePairs[numImages - 1].distance -
-        distanceImagePairs[0].distance
-    ) /
-    (numImages - 1);
+      vec3.sub(
+        positionVector,
+        referenceImagePositionPatient,
+        metadataForMiddleImage.imagePositionPatient
+      );
+      const distanceBetweenFirstAndMiddleImages = vec3.dot(
+        positionVector,
+        scanAxisNormal
+      );
+      zSpacing =
+        Math.abs(distanceBetweenFirstAndMiddleImages) /
+        Math.floor(imageIds.length / 2);
```

Comment:
why you are flooring here? I guess this result in inaccurate zspacing for odd numbers?

Reply:
That is just the way I am getting the middle imageId index so the average is calculated against the right number of images.